File: scala/src/main/scala/fundamentals/Conditionals.scala
package fundamentals

object Conditionals:
  def demonstrate(): Unit =
    println("=== Conditionals ===\n")
    
    // Basic if-else
    println("1. Basic if-else:")
    val number = 42
    if number > 0 then
      println(s"$number is positive")
    else if number < 0 then
      println(s"$number is negative")
    else
      println("The number is zero")
    
    // If expression (returns a value)
    println("\n2. If expression:")
    val max = if 10 > 5 then 10 else 5
    println(s"Maximum value: $max")
    
    // Multi-branch if-else
    println("\n3. Multi-branch if-else:")
    val grade = 85
    val letterGrade = if grade >= 90 then "A"
      else if grade >= 80 then "B"
      else if grade >= 70 then "C"
      else if grade >= 60 then "D"
      else "F"
    println(s"Grade $grade is letter grade $letterGrade")
    
    // Pattern matching (basic)
    println("\n4. Basic pattern matching:")
    val day = "Monday"
    val typeOfDay = day match
      case "Saturday" | "Sunday" => "Weekend"
      case _ => "Weekday"
    println(s"$day is a $typeOfDay")
    
    // Pattern matching with guards
    println("\n5. Pattern matching with guards:")
    val age = 25
    val category = age match
      case n if n < 13 => "Child"
      case n if n < 20 => "Teenager"
      case n if n < 30 => "Young Adult"
      case n if n < 50 => "Adult"
      case _ => "Senior"
    println(s"Age $age is categorized as: $category")
    
    // Pattern matching with case classes
    println("\n6. Pattern matching with case classes:")
    sealed trait Shape
    case class Circle(radius: Double) extends Shape
    case class Rectangle(width: Double, height: Double) extends Shape
    case class Triangle(base: Double, height: Double) extends Shape
    
    def calculateArea(shape: Shape): Double = shape match
      case Circle(r) => math.Pi * r * r
      case Rectangle(w, h) => w * h
      case Triangle(b, h) => 0.5 * b * h
    
    val shapes = List(
      Circle(5),
      Rectangle(4, 6),
      Triangle(3, 8)
    )
    
    shapes.foreach { shape =>
      val area = calculateArea(shape)
      println(s"Area of $shape is $area")
    }
    
    // Option handling with pattern matching
    println("\n7. Option handling:")
    def findDepartment(id: Int): Option[String] = id match
      case 1 => Some("HR")
      case 2 => Some("IT")
      case 3 => Some("Finance")
      case _ => None
    
    val deptId = 2
    val deptName = findDepartment(deptId) match
      case Some(name) => name
      case None => "Unknown Department"
    println(s"Department $deptId is: $deptName")
    
    // Type pattern matching
    println("\n8. Type pattern matching:")
    def describe(x: Any): String = x match
      case i: Int => s"Integer: $i"
      case s: String => s"String: $s"
      case d: Double => s"Double: $d"
      case l: List[_] => s"List with ${l.size} elements"
      case _ => "Unknown type"
    
    val values = List(42, "Hello", 3.14, List(1, 2, 3))
    values.foreach(value => println(describe(value)))
    
    // Pattern matching with tuples
    println("\n9. Pattern matching with tuples:")
    val point = (0, 0)
    val location = point match
      case (0, 0) => "Origin"
      case (0, y) => s"On Y-axis at y = $y"
      case (x, 0) => s"On X-axis at x = $x"
      case (x, y) => s"At point ($x, $y)"
    println(s"Point $point is: $location")
    
    println() // Empty line

File: kotlin/src/main/kotlin/fundamentals/Conditionals.kt
package fundamentals

sealed class Animal {
    abstract fun makeSound(): String
}

class Dog(val name: String) : Animal() {
    override fun makeSound() = "Woof!"
}

class Cat(val name: String) : Animal() {
    override fun makeSound() = "Meow!"
}

class Conditionals {
    fun demonstrate() {
        println("=== Conditionals ===\n")

        // Basic if expression
        println("1. Basic if expression:")
        val number = 42
        if (number > 0) {
            println("$number is positive")
        } else if (number < 0) {
            println("$number is negative")
        } else {
            println("$number is zero")
        }

        // If expression returning a value
        println("\n2. If expression as value:")
        val max = if (5 > 3) 5 else 3
        println("Max of 5 and 3 is $max")

        // Multi-line if expression
        val temperature = 25
        val description = if (temperature < 0) {
            "Very cold"
        } else if (temperature < 10) {
            "Cold"
        } else if (temperature < 20) {
            "Mild"
        } else if (temperature < 30) {
            "Warm"
        } else {
            "Hot"
        }
        println("\n3. Temperature is $temperature°C: $description")

        // When expression (switch equivalent)
        println("\n4. When expression:")
        val dayOfWeek = 3
        val day = when (dayOfWeek) {
            1 -> "Monday"
            2 -> "Tuesday"
            3 -> "Wednesday"
            4 -> "Thursday"
            5 -> "Friday"
            6 -> "Saturday"
            7 -> "Sunday"
            else -> "Invalid day"
        }
        println("Day $dayOfWeek is $day")

        // When with multiple values
        println("\n5. When with multiple values:")
        val number2 = 15
        when (number2) {
            0 -> println("Zero")
            1, 2, 3 -> println("Small number")
            in 4..9 -> println("Medium number")
            in 10..20 -> println("Large number")
            else -> println("Very large number")
        }

        // When without argument
        println("\n6. When without argument:")
        val x = 20
        val y = 10
        when {
            x > y -> println("x is greater than y")
            x < y -> println("x is less than y")
            else -> println("x equals y")
        }

        // Smart casts with when
        println("\n7. Smart casts with when:")
        val pet: Animal = Dog("Rex")
        when (pet) {
            is Dog -> println("${pet.name} says: ${pet.makeSound()}")
            is Cat -> println("${pet.name} says: ${pet.makeSound()}")
        }

        // When expression returning value
        println("\n8. When expression as value:")
        val obj: Any = "Hello"
        val result = when (obj) {
            is String -> "String of length ${obj.length}"
            is Int -> "Integer value $obj"
            is Boolean -> "Boolean value $obj"
            else -> "Unknown type"
        }
        println("Object type: $result")

        // Null safety with if
        println("\n9. Null safety with if:")
        val nullableValue: String? = null
        if (nullableValue != null) {
            println("Value is: $nullableValue")
        } else {
            println("Value is null")
        }

        // Combining conditions
        println("\n10. Combining conditions:")
        val age = 25
        val hasID = true
        if (age >= 18 && hasID) {
            println("Can enter the venue")
        } else {
            println("Cannot enter the venue")
        }

        // Range checks
        println("\n11. Range checks:")
        val score = 85
        when (score) {
            in 90..100 -> println("Excellent")
            in 80..89 -> println("Very Good")
            in 70..79 -> println("Good")
            in 60..69 -> println("Pass")
            else -> println("Fail")
        }

        println() // Empty line
    }
}